File: LinkedList/SinglyLinkedList/Basic/LastElementToFront.java
package Basic;

import x.*;

class LastElementToFront {
    public static void main(String[] args) {
        LinkedList l1 = new LinkedList();
        l1.append(1);
        l1.append(2);
        l1.append(3);
        l1.append(4);

        l1.print();
        System.out.println();
        lastToFront(l1);
        l1.print();
    }

    static void lastToFront(LinkedList list) {
        if (list.head == null || list.head.next == null) // list with 0 or 1 values
            return;

        // find last and prev nodes
        Node last = list.head, prev = list.head;
        while (last.next != null) {
            prev = last;
            last = last.next;
        }

        prev.next = null;
        last.next = list.head;
        list.head = last;
    }
}

File: LinkedList/SinglyLinkedList/Basic/LastElementToFront.js
// head -> 1 -> 2(prev) -> 3(cur) -> null;
// head -> 1 -> 2(prev) -> null  3(cur) -> head;
import { LinkedList } from "../x/LinkedList.js";

const list = new LinkedList();
list.append(1);
list.append(2);
list.append(3);
list.append(31);
list.append(98);

list.lastElementToFront = function() {
	// find the last element
	let cur = this.head,
		prev = null;

	while (cur.next != null) {
		prev = cur;
		cur = cur.next;
	}

	cur.next = this.head;
	this.head = cur;
	prev.next = null;
};

list.print();
list.lastElementToFront();
list.print();
